File: Data Structures & Algorithms/contains-duplicate-ii/submission-2.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        seen = set()
        left = 0

        for right in range(len(nums)):
            if abs(left - right) > k:
                seen.remove(nums[left])
                left += 1
            if nums[right] in seen:
                return True
            seen.add(nums[right])
        return False

File: Data Structures & Algorithms/sliding-window-maximum/submission-2.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int[] result = new int[nums.length - k + 1];
        Deque<Integer> deque = new ArrayDeque<>();
        int left = 0, right = 0;

        while (right < nums.length) {
            while (!deque.isEmpty() && nums[deque.getLast()] < nums[right]) {
                deque.removeLast();
            }
            deque.addLast(right);

            if (left > deque.getFirst()) {
                deque.removeFirst();
            }

            if (right + 1 >= k) {
                result[left] = nums[deque.getFirst()];
                left++;
            }
            right++;
        }
        return result;
    }
}
